File: python/final_nonfinal_tokens.py
from typing import Iterable
from soniox.transcribe_live import transcribe_stream
from soniox.speech_service import SpeechClient


def iter_audio() -> Iterable[bytes]:
    with open("../test_data/test_audio_long.flac", "rb") as fh:
        while True:
            audio = fh.read(1024)
            if len(audio) == 0:
                break
            yield audio


def main():
    with SpeechClient() as client:
        # All final tokens will be collected in this list.
        all_final_tokens = []

        for result in transcribe_stream(
            iter_audio(),
            client,
            model="en_v2_lowlatency",
            include_nonfinal=True,
        ):
            # Split current result response into final and non-final tokens.
            final_tokens = []
            nonfinal_tokens = []
            for word in result.words:
                if word.is_final:
                    final_tokens.append(word.text)
                else:
                    nonfinal_tokens.append(word.text)

            # Append current final tokens to all final tokens.
            all_final_tokens += final_tokens

            # Print all final tokens and current non-final tokens.
            all_final_tokens_str = "".join(all_final_tokens)
            nonfinal_tokens_str = "".join(nonfinal_tokens)
            print(f"Final: {all_final_tokens_str}")
            print(f"Non-final: {nonfinal_tokens_str}")
            print("-----")


if __name__ == "__main__":
    main()

File: speech_to_text_legacy/node/final_nonfinal_tokens.js
const fs = require("fs");
const { SpeechClient } = require("@soniox/soniox-node");

/**
 * Set your Soniox Cloud API key:
 * from command line: export SONIOX_API_KEY=<YOUR-API-KEY>
 * or
 * pass config object: new SpeechClient({ api_key: "<YOUR-API-KEY>" })
 */
const speechClient = new SpeechClient();

(async function () {
    // All final tokens will be collected in this list.
    let all_final_tokens = [];

    const onDataHandler = async (result) => {
        if (!result) {
            return;
        }

        const final_tokens = [];
        const nonfinal_tokens = [];

        for (word of result.words) {
            if (word.is_final) {
                final_tokens.push(word.text);
            } else {
                nonfinal_tokens.push(word.text);
            }
        }

        // Append current final tokens to all final tokens.
        all_final_tokens = all_final_tokens.concat(final_tokens);

        console.log(`Final: ${all_final_tokens.join("")}`);
        console.log(`Non-final: ${nonfinal_tokens.join("")}`);
        console.log("-----");
    };

    const onEndHandler = (error) => {
        if (error) {
            console.log(`Transcription error: ${error}`);
        }
    };

    // transcribeStream returns an object with ".writeAsync()" and ".end()"
    // methods - use them to send data and end stream when done.
    const stream = speechClient.transcribeStream(
        {
            model: "en_v2_lowlatency",
            include_nonfinal: true
        },
        onDataHandler,
        onEndHandler
    );

    // Open file as stream.
    const CHUNK_SIZE = 1024;
    const streamSource = fs.createReadStream(
        "../test_data/test_audio_long.flac",
        {
            highWaterMark: CHUNK_SIZE,
        }
    );

    // Simulate data streaming
    for await (const audioChunk of streamSource) {
        await stream.writeAsync(audioChunk);
    }
    stream.end();
})();
